[P092] New PlugIn to decode the DL-Bus

uwekaditz opened the pull request:
The DL-Bus is used in heating control units e.g. sold by Technische Alternative (www.ta.co.at).
The idea for this plug-in is based on Martin Kropf's project UVR31_RF24 (https://github.com/martinkropf/UVR31_RF24)
The plug-in is tested and workis fine for the ESR21 device.
The plug-in should be also able to decode the information from the UVR31, UVR1611 and UVR 61-3 devices but not tested yet.
The selected input needs a voltage divider as follows because the DL-Bus runs on 12 volts.

Grovkillen committed:
Merge pull request #1 from letscontrolit/mega

update from master

Grovkillen committed:
Merge pull request #2 from letscontrolit/mega

from master

TD-er committed:
[JSON] Fix issues with generated JSON 

Some JSON was invalid, for example when the closing entry of a list was optional (WiFi scanner)
The generated JSON for the I2C scanner was a bit useless, so made it more useful by only outputting addresses which have some device and included the list of supported devices.

uwekaditz committed:
[P092] New PlugIn to decode the DL-Bus

The DL-Bus is used in heating control units e.g. sold by Technische Alternative (www.ta.co.at).
The idea for this plug-in is based on Martin Kropf's project UVR31_RF24 (https://github.com/martinkropf/UVR31_RF24)
The plug-in is tested and workis fine for the ESR21 device.
The plug-in should be also able to decode the information from the UVR31, UVR1611 and UVR 61-3 devices but not tested yet.
The selected input needs a voltage divider as follows because the DL-Bus runs on 12 volts.

TD-er committed:
[JSON] Fix non valid timing stats json when no plugins defined (#2767)

On nodes without any plugins defined, the timing stats json would be invalid.

TD-er commented:
Can you also add basic documentation for it (see docs folder) ?

Also the plugin is not added to the define_plugin_sets.h file.

TD-er committed:
[JSON] Extend timing stats JSON end point with controller and misc stats

TD-er committed:
Merge pull request #2803 from TD-er/bugfix/json_missing_braces

[JSON] Fix issues with generated JSON

TD-er committed:
automatically updated release notes for mega-20191208

uwekaditz committed:
[P092] New PlugIn to decode the DL-Bus

Added to PLUGIN_SET_STABLE
Basic documentation added

uzi18 commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+  if (DataSettings.CurrentHmBytes > 2) {
+    byte b3 = ByteStream[power_index + 2];
+    byte b4 = ByteStream[power_index + 3];
+    high = 65536 * b4 + 256 * b3 + b2;
```
65536 is too much for `int high`

uwekaditz replied:
Thanks, changed to int32_t

TD-er replied:
int was already 32 bit, but making it `int32_t` is much more clear what it is as is shown here :)

uzi18 commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+      {
+        if (P092_init) {
+          addLog(LOG_LEVEL_ERROR, F("P092_init -> Already done!"));
+          return true;
```
what if someone change pin numer in configuration?

uwekaditz replied:
What do you mean?
P092_DLB_Pin is CONFIG_PIN1.
UserVar[event->BaseVarIndex] = NAN is used to show a NAN in the web device page.

uzi18 replied:
but if you return here and pin number is changed, plugin will not attachInterrupt to new pin

TD-er replied:
Yep, you have to restart the node to apply new settings with code like this.
The init is not cleared and you will enter this init call when you save the settings of the task.

uwekaditz replied:
Done.
A new init will be started if the device pin has changed.

uwekaditz committed:
Update _P092_DLbus.ino

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+  uint8_t mode;
+  float   value;
+} data_t;
+data_t ReadData;
```
Can you wrap these structs in a class or struct specific for this plugin?
Their names are so generic, they are bound to cause a conflict sooner or later with other code.
You have them global defined, so they are accessible by all other code.

uwekaditz replied:
Done!

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+        }
+        Plugin_092_SetIndices(PCONFIG(0));
+
+        if ((P092_ReceivedOK == false) || ((millis()-P092_LastReceived)>(Settings.TaskDeviceTimer[event->TaskIndex] * 1000/2))) {
```
We have functions for time computation.
For example `timePassedSince`.
Please use those functions, to make sure they are done right for the famous 49.7 days bug. (and improve readability)

uwekaditz replied:
Done!

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+#define MaxDatauint8_ts 64
+#define AdditionalRecuint8_ts 2
+#define MaxDataBits (((MaxDatauint8_ts + AdditionalRecuint8_ts) * (8 + 1 + 1) + 16) * 4)
+volatile uint8_t BitStream[MaxDataBits];           // received bit stream
```
The name suggests you store bits in there.
variables declared with `volatile` are stored in iram, which is quite limited on the ESP82xx.
Please make sure this is the most compact way to store the data.

Also the name of the variable is very generic and declared in the global scope.
Please prepend with P092, or wrap it in a struct.

uwekaditz replied:
All volatile variables are only used in the interrupt function.
It has to be fast, a bit change can occour every 2 milli seconds.
On most of the DL bus devices the bus clock is 488Hz.

uwekaditz replied:
I forgot one thing.
The time between the bit changes has to be measured quite accurate because the DL bus got a strange protocol:
Two eqal bits in sequence are without bit change, just the double length of the high signal. If there is another high bit following, a short low pulse is issued. The same for lows in sequence.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+  MaxDoublePulseWidth = DataSettings.MaxDoublePulseWidth;
+  addLog(LOG_LEVEL_INFO, F("P092_receiving ..."));
+  P092_receiving = true;
+  interrupts ();                                // interrupts allowed now, next instruction WILL be executed
```
Please move the `addLog` command past the command to enable the interrupts again.
This may take quite a lot of cycles to process and the log you try to output is static.

uwekaditz replied:
Done!

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+  P092_receiving = true;
+  interrupts ();                                // interrupts allowed now, next instruction WILL be executed
+  uint32_t start=millis();
+  while ((millis()-start)<100) {
```
Please use `timeOutReached` here  (or `timePassedSince`)
But maybe calling `delay(100)` is a bit easier ;)

Just wondering, why do you need to wait 100 msec?
Why not re-schedule the plugin to do the next stuff? 
For example 100 msec is the interval of 2 calls to 10-per-sec function.
Waiting for 100 msec is blocking at least 5 calls to 50/sec and a run of 10/sec.

uwekaditz replied:
I changed it, 100ms are just a timeout for receiving the first bit change.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+    return false;
+  }
+  while (P092_receiving) {
+    // wait for end of P092_receiving
```
How long do you have to wait here?
Waiting is bad, m'kay?

uwekaditz replied:
I added some comments about the looping time.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+
+void P092_Pin_changed(void) {
+  TimeInMicros = micros();
+  time_diff = TimeInMicros - last_bit_change;
```
Please use `timePassedSince` (there is also a usec version)

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+#define AUTO        0b00
+#define NORMAL      0b01
+#define LOWER       0b10
+#define STANDBY     0b11
```
These defines use very generic names.
I am really surprised if you would not get warnings about these being re-defined.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+    outputs = (uint8_tStream[DataSettings.IdxOutput + 1] << 8) | uint8_tStream[DataSettings.IdxOutput];
+  else
+    outputs = uint8_tStream[DataSettings.IdxOutput];
+  ReadData.value = (!!(outputs & (1 << (number - 1))));
```
Using `!!` is a nice trick to cast it to bool, but please add some comments here why you cast it to a bool like this.
The value is a float, so it is not really clear why you store it as a bool.

uwekaditz replied:
I changed it to an if-else-condition.

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+boolean P092_fetch_speed(int number) {
+  uint8_t speeduint8_t;
+
+  speeduint8_t = uint8_tStream[DataSettings.IdxDrehzahl + (number - 1)];
```
There is no check for the range of the index in the array.

uwekaditz replied:
Done!

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+boolean P092_fetch_analog(int number) {
+  uint8_t analoguint8_t;
+
+  analoguint8_t = uint8_tStream[DataSettings.IdxAnalog + (number - 1)];
```
No check for the range of the array index.

uwekaditz replied:
Done!

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+boolean P092_CheckHmRegister(int number) {
+  switch (number) {
+    case 1:
+      if ((uint8_tStream[DataSettings.IdxHmRegister] & 0x1) != 0x01)
```
These bitwise checks can be simpler. Just compare them with 0.
That also reduces the chance of type errors as you don't have to copy/paste the bitmask.

uwekaditz replied:
Done!

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -0,0 +1,1180 @@
+
+  if (P092_CheckHmRegister(number) == false)
+    return false;
+  uint8_t b1 = uint8_tStream[power_index];
```
`power_index` is not initialized when constructed and I am not sure it is set when this function is first called.

uwekaditz replied:
Done!

TD-er reviewed (changes requested):
Just a quick check of the code.

Grovkillen committed:
[README] updated with link to ICONS8

The new GUI will use icons from ICONS8 and since they are giving us a free premium account we thank them by linking to their web site.

Grovkillen committed:
[DOCS] P052 Senseair S11 added

Grovkillen committed:
[DOCS] S11 fixed some incorrect data

uwekaditz committed:
Update _P092_DLbus.ino

FIX: ReInit if device pin has changed
CHG: Unique names vor variables and defines
CHG: using usecPassedSince() and timePassedSince()
CHG: reduce waiting time in Plugin_092_Receiving(), it is just a timeout
CHG: added comments about looping time
CHG: check for the range of the index in the array in P092_fetch_speed() and P092_fetch_analog()

uzi18 committed:
LM75A - add multiple sensors support #2808

TD-er commented on src/_P092_DLbus.ino:
```diff
@@ -644,61 +672,66 @@ boolean Plugin_092_Receiving(void) {
+    // ESR21:     (((31 + 2) * (8 + 1 + 1) + 16) * 3) = 1038 bits @ 488Hz = 2,13s
+    // UVR31:     ((( 8 + 2) * (8 + 1 + 1) + 16) * 3) =  348 bits @  50Hz = 6,96s
+    // UVR1611:   (((64 + 2) * (8 + 1 + 1) + 16) * 3) = 2028 bits @ 488Hz = 4,16s
+    // UVR 61-3:  (((62 + 2) * (8 + 1 + 1) + 16) * 3) = 2028 bits @ 488Hz = 4,03s
```
Do you really need to actively wait here?
waiting for several seconds is a big no-no as you are blocking all other functionality of the node.
Please have a good look at what you are waiting for here.
The received bits will be processed in an interrupt-driven ISR call, so why wait here?
This can typically be handled in the 50/sec or 10/sec call and if a reading is finished you simply re-schedule the task to perform a PLUGIN_READ to collect the data.

Please have a look at the code of BME280.

uwekaditz replied:
I know, that this long blocking of all other functionalities is very bad, but I'm afraid that I loose some more bit change information while running other stuff as I already encounter.
**The if-condition on line 711 is already handling (that means guessing!) lost bit changes.**
So far the code runs fine on my WEMOS D1R1 mini for more than 1 year.
I just use two devices (DL bus and Sysinfo). The data of both devices are send via http to my FHEM home automation.
Nevertheless I'll try another code on my device (ESR21) coming weekend.
By the way I got some other questions:

- How big is the IRAM on the ESP82xx devices?

- Is there a priority for the interrupts?

- Can an interrupt function be interrupted (e.g. `attachInterrupt(digitalPinToInterrupt(P092_DLB_Pin), P092_Pin_changed, CHANGE)`)
by another interrupt?

TD-er replied:
> By the way I got some other questions:
> 
> * How big is the IRAM on the ESP82xx devices?
> * Is there a priority for the interrupts?
> * Can an interrupt function be interrupted (e.g. `attachInterrupt(digitalPinToInterrupt(P092_DLB_Pin), P092_Pin_changed, CHANGE)`)
>   by another interrupt?

The iRAM is quite small as in every now and then we run into issues where we cannot build our sketch.
I am not sure how big it actually is, but I do know there is not much left for our code to use.
It is also a bit confusing what iram actually is.
It is used to store executable code from flash, but some code (marked with the `ICACHE_RAM_ATTR`) and some variables (marked `volatile`) will remain pinned in that memory and never "swapped out" for other data or code.

Any way it is really important to have a good look at what to declare to remain pinned in there.


I don't know enough about priority on interrupts.
It is very well possible some interrupt does interfere with other interrupts.
I have seen some code specifically disabling serial interrupts for example, 
Maybe you have already seen this short intro on interrupts by random nerd tutorials: [ESP8266 Interrupts and Timers using Arduino IDE (NodeMCU)](https://randomnerdtutorials.com/interrupts-timers-esp8266-arduino-ide-nodemcu/)

You do call `yield()` in your loop.
The execution time of a yield is not constant and I can imagine it may interfere with the interrupt handling.
Maybe you can find some time in the decoding of the data where it is less of a problem if you miss a beat. For example between 2 bytes instead of 2 bits.

uwekaditz replied:
Thanks for your answers.
I only call `yield()` to keep the ESP at life (watchdag and so on). The excution time does not matter because the bit change detection is handled by the interrupt function.
The waiting loop is only used to avoid any distortion of the bit change receiving.
Nothing else is done while looping and waiting.
The decoding of the received bit changes (stored in `volatile uint8_t DLbus_BitStream[DLbus_MaxDataBits]`)  is already outside of the interrupt and does not have a wait loop.

TD-er replied:
> Nothing else is done while looping and waiting.

That's a bit of my point. It is not uncommon to have more sensors operational.

uwekaditz replied:
I'll try another code as soon as possible.
By the way what is the best variable definition for `uint8_t DLbus_BitStream[DLbus_MaxDataBits]` used in the interrupt function and outside of the interrupt. I'll try to access this array outside of the interrupt only while the interrupt is paused.

TD-er replied:
If you need it in an ISR function, then it must be declared volatile (or inside the `ICACHE_RAM_ATTR` function).
Not sure if declaring a pointer volatile and assigning a heap allocated vector to it is treated the same, although it should be.

Do you store bits in it, or do you compact it into bytes? (hopefully the last)
Please make sure not to allocate more than needed, so if the size of the array describes the number of compacted bits, then don't call it something referred to as bits.

TD-er replied:
Oh and if you call some function from the ISR call, then that function also needs to have the `ICACHE_RAM_ATTR`

uwekaditz replied:
I declared only variables needed in the interrupt function as `volatile`.
Do you have a code snippet for your proposal:
_declaring a pointer volatile and assigning a heap allocated vector to it_
No other function is called from the ISR call!
The received Bits are extendend to uint8_t, because they are extended by the timing flags needed for the analyze and processing functions.
I call it still BitStream but there is a comment to it.
`volatile uint8_t DLbus_BitStream[DLbus_MaxDataBits];  // received bit stream (each bit is extended to uint8_t, containing the timing flags)`, but it is also possible to use another name more related to bytes.

TD-er replied:
as long as the data is not stored as 1 bit per byte, then it is OK :)
Like I said, I only did have a very quick look at the code and only commented on what caught my eye immediately.
As I said, the name of the allocation sounded like a bit per byte storage, but if you also encode timing in it, then it sounds like good use of the array.

My remark about the extra functions was with a simple idea I had in mind, to pack the data in a separate function, but then it would have been called from within the ISR function and...
Well as always, my mind is running way faster than I can type, which makes things often a bit more confusing for the person reading my replies. (sorry 'bout that)

TD-er committed:
Merge pull request #2810 from uzi18/dallas

LM75A - add multiple sensors support #2808

Grovkillen committed:
[DOCS] S11 fixed PR after Gijs review

Grovkillen committed:
[DOCS] S11 re-phrased a sentance

TD-er committed:
[MQTT] Fix change in client.connected() behavior for pubsubclient

See this issue where the changes in behavior for wificlient.connected() is discussed: https://github.com/esp8266/Arduino/issues/6701

A change dating back to April 2018 ( https://github.com/esp8266/Arduino/pull/4626 ) does have some implications on how to interpret the client.connected() result.
It is very well possible there is still data in the buffers waiting to be read.
So the loop() function of pubsubclient should first try to empty the received data and then decide what to do with the connected state.

So reading is fine, if there is data available, but writing needs an extra check.

TD-er committed:
[WiFi] Improve reconnect management

Use of  connection failure threshold is now also trying to disconnect/reconnect a number of times before reboot.

TD-er committed:
Merge pull request #2809 from Grovkillen/mega

[DOCS] P052 Senseair S11 added

TD-er committed:
Merge pull request #2811 from TD-er/bugfix/pubsubclient_connected

[MQTT] Fix change in client.connected() behavior for pubsubclient

uwekaditz committed:
Merge branch 'PullRequest' of https://github.com/uwekaditz/ESPEasy into PullRequest

TD-er commented:
Hmm, it looks like you merged in all commits on the mega branch.
What you should have done is to rebase your branch onto the HEAD of the mega branch and then force push the branch.
Only you can fix this.

uwekaditz commented:
How to fix my failure?

TD-er commented:
The simplest is probably to create a new branch (at HEAD of mega) and cherry pick each of your commits from this branch you PR'ed.

Then open a new PR with that new branch.

There are other options, but I think they are harder to explain.

uwekaditz commented:
Moved to PR #2815 
This PR can be closed!
